Source/WebCore/css/ShorthandSerializer.cpp

```diff
@@ -0,0 +1,1175 @@
+        return serializeFontSynthesis();
+    case CSSPropertyGrid:
+        return serializeGrid();
+    case CSSPropertyGridArea:
+        return serializeGridArea();
+    case CSSPropertyGridColumn:
+    case CSSPropertyGridRow:
+        return serializeGridRowColumn();
+    case CSSPropertyGridTemplate:
+        return serializeGridTemplate();
+    case CSSPropertyMarker:
+        return serializeCommonValue();
+    case CSSPropertyOffset:
+        return serializeOffset();
+    case CSSPropertyPageBreakAfter:
+    case CSSPropertyPageBreakBefore:
+        return serializePageBreak();
+    case CSSPropertyPageBreakInside:
+    case CSSPropertyWebkitColumnBreakInside:
+        return serializeBreakInside();
+    case CSSPropertyTextDecorationSkip:
+    case CSSPropertyTextDecoration:
+    case CSSPropertyWebkitBackgroundSize:
+    case CSSPropertyWebkitPerspective:
+    case CSSPropertyWebkitTextOrientation:
+        return serializeLonghandValue(0);
+    case CSSPropertyTransformOrigin:
+        return serializeLonghandsOmittingTrailingInitialValue();
+    case CSSPropertyWebkitColumnBreakAfter:
+    case CSSPropertyWebkitColumnBreakBefore:
+        return serializeColumnBreak();
+    default:
+        ASSERT_NOT_REACHED();
+        return String();
+    }
+}
+
+String ShorthandSerializer::serializeLonghands() const
+{
+    return serializeLonghands(length());
+}
+
+String ShorthandSerializer::serializeLonghands(unsigned lengthLimit, const char* separator) const
+{
+    ASSERT(lengthLimit <= length());
+    switch (lengthLimit) {
+    case 1:
+        return serializeLonghandValue(0);
+    case 2:
+        return makeString(serializeLonghandValue(0), separator, serializeLonghandValue(1));
+    case 3:
+        return makeString(serializeLonghandValue(0), separator, serializeLonghandValue(1), separator, serializeLonghandValue(2));
+    case 4:
+        return makeString(serializeLonghandValue(0), separator, serializeLonghandValue(1), separator, serializeLonghandValue(2), separator, serializeLonghandValue(3));
+    default:
+        StringBuilder result;
+        auto prefix = "";
+        for (unsigned i = 0; i < lengthLimit; ++i)
+            result.append(std::exchange(prefix, separator), serializeLonghandValue(i));
+        return result.toString();
+    }
+}
+
+String ShorthandSerializer::serializeLonghandsOmittingInitialValues() const
+{
+    StringBuilder result;
+    auto prefix = "";
+    for (auto longhand : longhands()) {
+        if (!isInitialValue(longhand))
+            result.append(std::exchange(prefix, " "), serializeValue(longhand));
+    }
+    return result.isEmpty() ? serializeLonghandValue(0) : result.toString();
+}
+
+String ShorthandSerializer::serializeLonghandsOmittingTrailingInitialValue(const char* separator) const
+{
+    ASSERT(length() >= 2);
+    return serializeLonghands(length() - isLonghandInitialValue(length() - 1), separator);
+}
+
+inline String ShorthandSerializer::serializeCommonValue() const
```

Comment:
Does this `inline` do anything?

Reply:
It does not. I gave in to outdated urges and put in quite a few `inline` like this one that likely have no effect, but my meaning was "I hope this gets inlined".